### PR TITLE
Use cuGetProcAddress to load CUDA entry points

### PR DESCRIPTION
**Pull request by trevor-m**
Sometimes the v2 APIs can have different function signatures which will lead to errors if it doesn't match what we use in the iree code. V2 lookup was originally added here https://github.com/antiagainst/iree/commit/9f42d0448af7f80700cd613d69f8f4cc67583e1b

NVIDIA recommends using `cuGetProcAddress` to avoid version mixing which can otherwise cause undefined behavior: https://docs.nvidia.com/cuda/cuda-driver-api/version-mixing-rules.html#version-mixing-rules

### CODE REVIEW

**Review by benvanik — APPROVED** [review]

**Comment by antiagainst** [comment]
Hi @trevor-m would you mind to mirror the same change under `experimental/cuda2/`? Should be straightforward if this change passes CI.

**Review by antiagainst — APPROVED** [review]

**Comment by trevor-m** [comment]
Getting `CUDA_ERROR_INVALID_CONTEXT` from https://github.com/openxla/iree/blob/main/runtime/src/iree/hal/drivers/cuda/cuda_allocator.c#L361, taking a look

**Comment by trevor-m** [comment]
Looks like we need `cuMemAlloc_v2` to fix the `CUDA_ERROR_INVALID_CONTEXT`.

However, since cuda.h defines cuMemAlloc as cuMemAlloc_v2, it makes things a little weird and we have to do
```
#undef cuMemAlloc
CU_PFN_DECL(cuMemAlloc, CUdeviceptr*, size_t)
CU_PFN_DECL(cuMemAlloc_v2, CUdeviceptr*, size_t)
```

If we used https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1gb82d2a09844a58dd9e744dc31e8aa467 we would be safe knowing that all cuda APIs will be from the same version and be compatible with eachother. Let me give it a try.

File: experimental/cuda2/cuda_dynamic_symbols.c [hunk 1]
```diff
@@ -20,21 +20,25 @@ static const char* iree_hal_cuda_dylib_names[] = {
...
-    IREE_RETURN_IF_ERROR(iree_dynamic_library_lookup_symbol(            \
-        syms->dylib, name, (void**)&syms->cuda_symbol_name));           \
-    static const char* name_v2 = IREE_CONCAT(#cuda_symbol_name, "_v2"); \
-    void* fptr_v2;                                                      \
-    iree_dynamic_library_lookup_symbol(syms->dylib, name_v2, &fptr_v2); \
-    if (fptr_v2) syms->cuda_symbol_name = fptr_v2;                      \
+// Load CUDA entry points.
+static iree_status_t iree_hal_cuda_dynamic_symbols_resolve_all(
+    iree_hal_cuda_dynamic_symbols_t* syms) {
+  CUresult (*cuGetProcAddress)(const char*, void**, int, cuuint64_t);
```
Comment (benvanik) [inline experimental/cuda2/cuda_dynamic_symbols.c]: you can put this in the dynamic_symbol_tables.h and do the lookup into it (&syms->cuGetProcAddress) - then the code below can use the IREE_CUDA_RETURN_IF_ERROR macro that handles converting CUresults to statuses with nice messages

so something like `IREE_CUDA_RETURN_IF_ERROR(syms, cuGetProcAddress(...), "resolving " # cuda_symbol_name);`

**Review by benvanik — CHANGES_REQUESTED** [review]

File: experimental/cuda2/cuda_dynamic_symbols.c (hunk 1 above)
Comment (benvanik) [reply experimental/cuda2/cuda_dynamic_symbols.c]: (advantage of this being that plugins/etc can get the symbol for querying their own symbols that we don't include in the base table)

Comment (trevor-m) [reply experimental/cuda2/cuda_dynamic_symbols.c]: Sounds good, thanks! Should we move `cuGetErrorName`/`cuGetErrorString` to the top of the table so they are loaded first incase they are needed by `IREE_CUDA_RETURN_IF_ERROR` to decode any errors?

Comment (trevor-m) [reply experimental/cuda2/cuda_dynamic_symbols.c:30]: It seems `CUDA_RETURN_IF_ERROR` is eating the `__VA_ARGS__` when used inside `CU_PFN_DECL` and causing some weird compilation errors. Any advice? @benvanik

Comment (trevor-m) [reply experimental/cuda2/cuda_dynamic_symbols.c]: @benvanik I took a stab at it. Let me know if you think of a cleaner way to handle the error checking.

File: experimental/cuda2/cuda_dynamic_symbols.c [hunk 2]
```diff
@@ -20,21 +20,28 @@ static const char* iree_hal_cuda_dylib_names[] = {
...
+  // Since cuGetProcAddress is in the symbol table, it will be loaded again
+  // through cuGetProcAddress. cuGetProcAddress_v2 is added in CUDA 12.0 and has
+  // a new function signature. If IREE_CUDA_DRIVER_API_VERSION is increased to
+  // >=12.0, then make sure we are using the correct signature.
+  IREE_RETURN_IF_ERROR(iree_dynamic_library_lookup_symbol(
+      syms->cuda_library, "cuGetProcAddress", (void**)&syms->cuGetProcAddress));
+#define CU_PFN_DECL(cudaSymbolName, ...)                       \
+  {                                                            \
+    static const char* kName = #cudaSymbolName;                \
+    CUDA_RETURN_IF_ERROR(                                      \
```
Comment (antiagainst) [inline experimental/cuda2/cuda_dynamic_symbols.c]: This should be `IREE_CUDA_RETURN_IF_ERROR`. (We cleaned up various macros and have proper prefixes for them in cuda2.)

File: experimental/cuda2/cuda_dynamic_symbols.c [hunk 3]
```diff
@@ -20,21 +20,28 @@ static const char* iree_hal_cuda_dylib_names[] = {
...
+// Load CUDA entry points.
+static iree_status_t iree_hal_cuda_dynamic_symbols_resolve_all(
+    iree_hal_cuda_dynamic_symbols_t* syms) {
+  // Since cuGetProcAddress is in the symbol table, it will be loaded again
+  // through cuGetProcAddress. cuGetProcAddress_v2 is added in CUDA 12.0 and has
+  // a new function signature. If IREE_CUDA_DRIVER_API_VERSION is increased to
+  // >=12.0, then make sure we are using the correct signature.
+  IREE_RETURN_IF_ERROR(iree_dynamic_library_lookup_symbol(
+      syms->cuda_library, "cuGetProcAddress", (void**)&syms->cuGetProcAddress));
+#define CU_PFN_DECL(cudaSymbolName, ...)                       \
```
Comment (antiagainst) [inline experimental/cuda2/cuda_dynamic_symbols.c]: `cuda_symbol_name` to follow coding conventions?

File: experimental/cuda2/cuda_dynamic_symbols.c [hunk 4]
```diff
@@ -20,21 +20,28 @@ static const char* iree_hal_cuda_dylib_names[] = {
...
+  {                                                            \
+    static const char* kName = #cudaSymbolName;                \
+    CUDA_RETURN_IF_ERROR(                                      \
+        syms,                                                  \
+        cuGetProcAddress(kName, (void**)&syms->cudaSymbolName, \
+                         IREE_CUDA_DRIVER_API_VERSION,         \
+                         CU_GET_PROC_ADDRESS_DEFAULT),         \
+        "cuGetProcAddress");                                   \
```
Comment (antiagainst) [inline experimental/cuda2/cuda_dynamic_symbols.c]: "when resolving " #cuda_symbol_name " using cuGetProcAddress" ?

**Review by antiagainst — CHANGES_REQUESTED** [review]
Could you add `-DIREE_EXTERNAL_HAL_DRIVERS=cuda2` to make sure after the modification it still builds? Thanks! :)

**Comment by trevor-m** [comment]
Thanks @antiagainst, I have now fixed the cuda2 build issues and confirmed it works with `-DIREE_EXTERNAL_HAL_DRIVERS=cuda2`

File: experimental/cuda2/cuda_dynamic_symbols.c (hunk 1 above)
Comment (benvanik) [reply experimental/cuda2/cuda_dynamic_symbols.c]: Huh, hadn't considered that! What you have looks reasonable as I doubt those functions are going to change much/ever be missing.

**Review by benvanik — APPROVED** [review]
lgtm if @antiagainst is happy! thanks for this!

**Review by antiagainst — APPROVED** [review]

**Comment by antiagainst** [comment]
Thanks a lot!

**Comment by trevor-m** [comment]
@benvanik Can this be merged now?

**Pull request merged**